src/builder/value_parser.rs

```diff
@@ -1051,6 +1079,136 @@ where
     }
 }
 
+#[derive(Copy, Clone, Debug)]
+pub struct InRange<P>
+where
+    P: TypedValueParser,
+    P::Value: PartialOrd,
+{
+    parser: P,
+    bounds: (std::ops::Bound<P::Value>, std::ops::Bound<P::Value>),
+}
+
+impl<P: TypedValueParser> InRange<P>
+where
+    P::Value: PartialOrd + Clone,
+{
+    pub fn new<T, R: RangeBounds<P::Value>>(range: R) -> Self
+    where
+        T: ValueParserFactory<Parser = P>,
+    {
+        Self::with_parser(T::value_parser(), range)
+    }
+
+    pub fn with_parser<R: RangeBounds<P::Value>>(parser: P, range: R) -> Self {
+        Self {
+            parser,
+            bounds: (range.start_bound().cloned(), range.end_bound().cloned()),
+        }
+    }
+
+    pub fn in_range<R: RangeBounds<P::Value>>(self, range: R) -> InRange<P>
```

Comment:
The current behavior is if you do `value_parser!(u16).range(1..)`
- It gets parsed as a i64 internally, causing more parse errors to be treated as range errors
- The ranges compound with each other so you get `1..=u16::MAX`

With the current implementation,
- It doesn't allow narrowing the value's type, causing a regression in error messages
- It doesn't merge the ranges

Reply:
hmm. For the second part, that could potentially be resolved by require `T` to implement a trait that exposes the maximum and minimum values. Although that would require the type to actually have a minimum and maximum value (or maybe it returns an `Option`?), and of course implement the trait for any meaningful data types, such as the integral and floating point types in the standard library.

The first part is more difficult though. If we wanted to restrict this to integers, then possibly the current implementation could be changed to use i128 and u128 on platforms that support it. But i'm not sure there is any garantee that those will be at least as big as isize and usize. It also doesn't allow specifying a range for f32 or f64. Nor would it really be able to support custom types that have an ordering (although I'm not sure how much of a need there is for that).

Reply:
My original attempt at this had generic types for the parse type and the value type and required `TryInto` to be supported between them.